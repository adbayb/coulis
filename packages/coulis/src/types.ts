import {
	AdvancedPseudos,
	HtmlAttributes,
	PropertiesFallback,
	SimplePseudos,
} from "csstype";

// @note: UngreedyString is a special string type allowing literal enums getting widened to the super type string when specified
// It allows to enable string type with literal enums without loosing autocomplete DX
// eslint-disable-next-line @typescript-eslint/ban-types
type UngreedyString = string & {};

type Property = PropertiesFallback<number | UngreedyString>;

type AttributeSelectors = HtmlAttributes;

type PseudoSelectors = SimplePseudos | AdvancedPseudos;

export type StyleObject = {
	[Key in keyof Property]:
		| Property[Key]
		| Partial<
				Record<
					| "default"
					| PseudoSelectors
					| AttributeSelectors
					| UngreedyString,
					Property[Key]
				>
		  >;
};
