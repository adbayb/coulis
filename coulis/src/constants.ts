// eslint-disable-next-line unicorn/prefer-global-this
export const IS_BROWSER_ENVIRONMENT = typeof window !== "undefined";
