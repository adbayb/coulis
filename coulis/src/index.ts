export { createKeyframes } from "./modules/keyframes";
export { createCustomProperties } from "./modules/properties";
export { extractStyles } from "./modules/server";
export { createStyles, globalStyles } from "./modules/styles";
export { createVariants } from "./modules/variants";
