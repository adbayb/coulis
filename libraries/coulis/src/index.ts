export { createKeyframes } from "./modules/keyframes";
export { createCustomProperties } from "./modules/properties";
export { extract } from "./modules/server";
export { styles, createStyles, globalStyles } from "./modules/styles";
export { compose } from "./modules/tools";
export { createVariants } from "./modules/variants";
// export * from "./modules/stylesNext";
