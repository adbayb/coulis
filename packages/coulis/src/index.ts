// @todo: benchmarks
// - if multiple insert of styles per atomic css declaration is more or less efficient than one single style element ?
// @todo: hydrate createProcessor cache client side from data-coulis-keys tag ?
// @todo: rename `css` api to `styles` (universal platform meaning) -> plural since it can return multiple classnames and consistent with keyframes api
// @todo: rename `raw` api to `globals`? Keep keyframes
export { createCss, css, extractStyles, keyframes, raw } from "./api";
