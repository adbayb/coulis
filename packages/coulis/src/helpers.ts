import { UNITLESS_PROPERTIES } from "./constants";

export const hash = (str: string) => {
	// hash content based with FNV-1a algorithm:
	const FNVOffsetBasis = 2166136261;
	const FNVPrime = 16777619;
	let hash = FNVOffsetBasis;

	for (let i = 0; i < str.length; i++) {
		hash ^= str.charCodeAt(i);
		hash *= FNVPrime;
	}

	// @note: we convert hashed value to 32-bit unsigned integer
	// via logical unsigned shift operator >>>
	const uHash = hash >>> 0;

	// @note: a coulis className is generated by prefixing with "c"
	// and converting generated hash to hexadecimal
	return "c" + Number(uHash).toString(16);
};

export const isNumber = (value: unknown): value is number => {
	return typeof value === "number" || !Number.isNaN(Number(value));
};

export const isObject = (value: unknown): value is Record<string, unknown> => {
	return value !== null && typeof value === "object";
};

export const minify = (value: string) => {
	return value.replace(/\s{2,}|\s+(?={)|\r?\n/gm, "");
};

export const toDeclaration = (property: string, value: unknown) => {
	// @note: Anatomy of a css syntax:
	// .className { background-color: blue; color: red } = css rule-set
	// .className = selector
	// { background-color: blue; color: red } = declaration block (contains one or more declarations separated by semicolons)
	// background-color: blue = a declaration
	// background-color = property (or property name)
	// blue = value (or property value)

	// @section: from JS camelCase to CSS kebeb-case
	const normalizedProperty = property.replace(
		/([A-Z])/g,
		(matched) => `-${matched.toLowerCase()}`
	);
	// @section: format value to follow CSS specs (unitless number)
	const normalizedValue =
		typeof value !== "number" || UNITLESS_PROPERTIES[property]
			? value
			: `${value}px`;

	return `${normalizedProperty}:${normalizedValue};`;
};

export const toManyDeclaration = <StyleObject extends Record<string, unknown>>(
	styleObject: StyleObject
) => {
	let declarationBlock = "";

	for (const property of Object.keys(styleObject)) {
		const value = styleObject[property];

		declarationBlock += toDeclaration(property, value);
	}

	return declarationBlock;
};
