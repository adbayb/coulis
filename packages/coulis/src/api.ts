import { NO_CLASSNAME, SHORTHAND_PROPERTIES } from "./constants";
import { isNumber, isObject, minify, toDeclaration } from "./helpers";
import { StyleSheetType, createStyleSheet } from "./entities/stylesheet";
import { createCache } from "./entities/cache";
import { process } from "./entities/processor";
import {
	AtConditionalGroupingRule,
	AtTextualRule,
	AtomicStyleObject,
	GlobalStyleObject,
	KeyframeStyleObject,
} from "./types";

const styleSheet = createStyleSheet();
const cache = createCache(styleSheet);

/**
 * Create a contextual `atoms` tied to a [CSSGroupingRule](https://developer.mozilla.org/en-US/docs/Web/API/CSSGroupingRule)
 * (ie. an at-rule that contains other rules nested within it (such as @media, @supports conditional rules...))
 * @param groupRule The styling rule instruction (such as `@media (min-width: 0px)`)
 * @returns The contextual `atoms` helper
 */
export const createAtoms = (
	atRule: AtConditionalGroupingRule,
	condition: string
) => {
	return createAtomsFactory(`${atRule} ${condition}`);
};

export const atoms = (styleObject: AtomicStyleObject) => {
	return createAtomsFactory()(styleObject);
};

export const extractStyles = () => {
	const styleSheetTypes = Object.keys(styleSheet) as StyleSheetType[];
	const cacheEntries = cache.entries();
	const cacheKeys = Object.keys(cacheEntries);

	return styleSheetTypes.map((type) => {
		const style = styleSheet[type];
		const cssValue = minify(style.get());
		const keys = cacheKeys.filter((key) => cacheEntries[key] === type);

		return {
			keys,
			content: cssValue,
			type,
			toString() {
				return `<style data-type="${type}" data-keys="${keys.join(
					","
				)}" data-coulis>${cssValue}</style>`;
			},
		};
	});
};

export const globals = (styleObject: GlobalStyleObject) =>
	process({
		cache,
		key: JSON.stringify(styleObject),
		styleSheet: styleSheet.global,
		strategy() {
			let ruleSet = "";

			for (const selector of Object.keys(styleObject)) {
				const style = styleObject[selector as AtTextualRule];

				if (!style) continue;

				if (typeof style === "string") {
					ruleSet += `${selector} ${style};`;
				} else {
					let declarationBlock = "";

					for (const property of Object.keys(style)) {
						const value = style[property];

						declarationBlock += `${toDeclaration(
							property,
							value
						)};`;
					}

					ruleSet += `${selector}{${declarationBlock}}`;
				}
			}

			return ruleSet;
		},
	});

export const keyframes = (styleObject: KeyframeStyleObject) =>
	process({
		cache,
		key: JSON.stringify(styleObject),
		styleSheet: styleSheet.global,
		strategy({ className }) {
			let ruleSet = "";
			const selectors = Object.keys(styleObject) as Array<
				keyof typeof styleObject
			>;

			for (const selector of selectors) {
				const style = styleObject[selector];

				if (!style) continue;

				let declarationBlock = "";

				for (const property of Object.keys(style)) {
					const value = style[property];

					declarationBlock += `${toDeclaration(property, value)};`;
				}

				ruleSet += `${
					isNumber(selector) ? `${selector}%` : selector
				}{${declarationBlock}}`;
			}

			return `@keyframes ${className}{${ruleSet}}`;
		},
	});

const createAtomsFactory = (groupingRule = "") => {
	const wrapRuleSet = (ruleSet: string) => {
		return !groupingRule ? ruleSet : `${groupingRule}{${ruleSet}}`;
	};

	// eslint-disable-next-line sonarjs/cognitive-complexity
	return (styleObject: AtomicStyleObject) => {
		let classNames = "";

		for (const property of Object.keys(styleObject)) {
			const value = styleObject[property];
			const style = groupingRule
				? styleSheet.conditional
				: SHORTHAND_PROPERTIES[property]
				? styleSheet.shorthand
				: styleSheet.longhand;

			if (isObject(value)) {
				for (const selectorProperty of Object.keys(value)) {
					const selectorValue = value[selectorProperty];
					const isDefaultProperty = selectorProperty === "default";
					const className = process({
						cache,
						key: `${groupingRule}${property}${selectorValue}${
							isDefaultProperty ? "" : selectorProperty
						}`,
						styleSheet: style,
						strategy({ className }) {
							if (selectorValue === undefined)
								return NO_CLASSNAME;

							const declarationBlock = toDeclaration(
								property,
								selectorValue
							);

							return wrapRuleSet(
								`.${className}${
									isDefaultProperty ? "" : selectorProperty
								}{${declarationBlock}}`
							);
						},
					});

					if (className !== NO_CLASSNAME) {
						classNames = `${classNames} ${className}`;
					}
				}
			} else {
				const className = process({
					cache,
					key: `${groupingRule}${property}${value}`,
					styleSheet: style,
					strategy({ className }) {
						if (value === undefined) return NO_CLASSNAME;

						const declarationBlock = toDeclaration(property, value);

						return wrapRuleSet(
							`.${className}{${declarationBlock}}`
						);
					},
				});

				if (className !== NO_CLASSNAME) {
					classNames = `${classNames} ${className}`;
				}
			}
		}

		return classNames.trim();
	};
};
