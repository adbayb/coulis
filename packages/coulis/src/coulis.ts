import { SHORTHAND_PROPERTIES } from "./constants";
import { hash, isObject, merge, minify } from "./helpers";
import { StyleSheetKey, createStyleSheets } from "./domains/stylesheet";
import { createCache } from "./domains/cache";
import { createProcessor, toClassName } from "./domains/processor";
import { DeclarationBlock } from "./types";

const styleSheets = createStyleSheets();
const cache = createCache(styleSheets);
const processStyle = createProcessor(cache);

export const createCss = (groupRule: string) => {
	const formatRuleSetWithScope = (ruleSet: string) => {
		return !groupRule ? ruleSet : `${groupRule}{${ruleSet}}`;
	};

	return (...cssBlocks: DeclarationBlock[]) => {
		const cssBlock =
			cssBlocks.length <= 1 ? cssBlocks[0] : merge({}, ...cssBlocks);
		let classNames = "";

		for (const property in cssBlock) {
			const value = cssBlock[property];
			const styleSheet = groupRule
				? styleSheets.conditional
				: SHORTHAND_PROPERTIES[property]
				? styleSheets.shorthand
				: styleSheets.longhand;

			if (isObject(value)) {
				for (const selectorProperty in value) {
					const selectorValue =
						value[selectorProperty as keyof typeof value];
					const isDefaultProperty = selectorProperty === "default";
					const className = processStyle(
						isDefaultProperty
							? `${groupRule}${property}${selectorValue}`
							: `${groupRule}${property}${selectorValue}${selectorProperty}`,
						property,
						selectorValue,
						(className, declaration) =>
							formatRuleSetWithScope(
								`.${className}${
									isDefaultProperty ? "" : selectorProperty
								}{${declaration}}`
							),
						styleSheet
					);

					if (className !== null) {
						classNames = `${classNames} ${className}`;
					}
				}
			} else {
				const className = processStyle(
					`${groupRule}${property}${value}`,
					property,
					value,
					(className, declaration) =>
						formatRuleSetWithScope(`.${className}{${declaration}}`),
					styleSheet
				);

				if (className !== null) {
					classNames = `${classNames} ${className}`;
				}
			}
		}

		return classNames.trim();
	};
};

export const css = createCss("");

export const keyframes = (value: string) => {
	const key = hash(value);
	const className = `${toClassName(key)}`;

	if (cache.has(key)) {
		return className;
	}

	const declarationBlock = `@keyframes ${className}{${minify(value)}}`;

	styleSheets.global.commit(declarationBlock);
	cache.set(key, "global");

	return className;
};

export const extractStyles = () => {
	const styleSheetTypes = Object.keys(styleSheets) as StyleSheetKey[];
	const cacheEntries = cache.entries();
	const cacheKeys = Object.keys(cacheEntries);

	return styleSheetTypes.map((type) => {
		const styleSheet = styleSheets[type];
		const cssValue = minify(styleSheet.get());
		const keys = cacheKeys.filter((key) => cacheEntries[key] === type);

		return {
			keys,
			content: cssValue,
			type,
			toString() {
				return `<style data-coulis-type="${type}" data-coulis-keys="${keys.join(
					","
				)}">${cssValue}</style>`;
			},
		};
	});
};

export const raw = (value: string) => {
	const key = hash(value);

	if (cache.has(key)) {
		return;
	}

	styleSheets.global.commit(minify(value));
	cache.set(key, "global");
};
