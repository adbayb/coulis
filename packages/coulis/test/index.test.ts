import { createCss, css, extractStyles, keyframes, raw } from "../src";

test("should extract styles", () => {
	const cssMedia = createCss("@media (min-width: 400px)");

	raw(`html, body {
		background-color: lightcoral;
	}

	div {
		color: red;
	}`);

	const animationName = keyframes(`
from {
	transform: scale(1);
}

to {
	transform: scale(1.015);
}
`);

	const classNames = css({
		color: "lightblue",
		backgroundColor: {
			default: "lightcoral",
			":hover": "lightcyan",
			"[alt]": "lightgray",
		},
		padding: 10,
	});

	const mediaClassNames = cssMedia({
		color: "lightblue",
		backgroundColor: {
			default: "lightcoral",
			":hover": "lightcyan",
			"[alt]": "lightgray",
		},
		padding: 10,
	});

	expect(animationName).toBe("c68c11a40");
	expect(classNames).toBe(
		"c3c5816c4 cbb3db274 c571db958 ce9403afc cde6daf3c"
	);
	expect(mediaClassNames).toBe(
		"cfdaa40cc c83a26eac c8da3c910 c1225fd44 c9cc8640"
	);
	expect(extractStyles()).toMatchSnapshot();
});
