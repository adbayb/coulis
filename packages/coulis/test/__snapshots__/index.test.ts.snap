// Vitest Snapshot v1

exports[`should extract styles 1`] = `
[
  {
    "content": "@charset \\"utf-8\\";@import url('https://fonts.googleapis.com/css?family=Open+Sans&display=swap');html{box-sizing:border-box;}html,body{padding:0px;margin:0px;background-color:lightcoral;font-family:Open Sans, AliasedHelvetica;}*,*::before,*::after{box-sizing:inherit;}.globalClass+.otherGlobalClass{border:1px solid black;border-radius:4px;}@font-face{font-family:'AliasedHelvetica';src:local(Helvetica);}@keyframes c62e63a97{25%{transform:scale(1.25);}from{transform:scale(1);}50%{transform:scale(1.5);}to{transform:scale(1);}}",
    "keys": [
      "61d92872",
      "62e63a97",
    ],
    "toString": [Function],
    "type": "global",
  },
  {
    "content": ".c3c5816c4{color:lightblue;}.cbb3db274{background-color:lightcoral;}.c571db958:hover{background-color:lightcyan;}.ce9403afc[alt]{background-color:lightgray;}",
    "keys": [
      "3c5816c4",
      "bb3db274",
      "571db958",
      "e9403afc",
    ],
    "toString": [Function],
    "type": "longhand",
  },
  {
    "content": ".cde6daf3c{padding:10px;}",
    "keys": [
      "de6daf3c",
    ],
    "toString": [Function],
    "type": "shorthand",
  },
  {
    "content": "@media (min-width: 576px){.c721688f8{color:lightblue;}}@media (min-width: 576px){.c3d523ba0{background-color:lightcoral;}}@media (min-width: 576px){.cf33f0fbe:hover{background-color:lightcyan;}}@media (min-width: 576px){.cec510c6c[alt]{background-color:lightgray;}}@media (min-width: 576px){.ca8209d58{padding:10px;}}",
    "keys": [
      "721688f8",
      "3d523ba0",
      "f33f0fbe",
      "ec510c6c",
      "a8209d58",
    ],
    "toString": [Function],
    "type": "conditional",
  },
]
`;
